src/delta/delta.c

```diff
@@ -207,102 +208,99 @@ static int found_override(const char *top, const char *bottom) {
         return k;
 }
 
-static int enumerate_dir_d(Hashmap *top, Hashmap *bottom, Hashmap *drops, const char *toppath, const char *drop) {
+static int enumerate_dir_d(Hashmap *bottom, const char *toppath, const char *drop) {
         _cleanup_free_ char *unit = NULL;
-        _cleanup_free_ char *path = NULL;
         _cleanup_strv_free_ char **list = NULL;
-        char **file;
-        char *c;
+        char **file, *c, *path;
         int r;
 
         assert(!endswith(drop, "/"));
 
-        path = strjoin(toppath, "/", drop, NULL);
-        if (!path)
-                return -ENOMEM;
+        log_debug("Looking at %s", toppath);
 
-        log_debug("Looking at %s", path);
-
-        unit = strdup(drop);
-        if (!unit)
+        path = strdup(toppath);
+        if (!path)
                 return -ENOMEM;
 
-        c = strrchr(unit, '.');
+        c = strrchr(path, '.');
         if (!c)
```

Comment:
Our usual approach is to set variables to `NULL` in success case when a reference is taken down the line and we cannot free them ourselves anymore. This makes the 'ownership' of a chunk of memory clearer. It's not a big deal, but the repetition of the exact same cleanup lines smells like something that can be solved in a nicer way.

Also, even if original code was written in a strange way, that doesn't mean we can't clean up such locations when we revisit the code.